### ISSUE TITLE
High memory usage when resizing thumbnails

### ISSUE THREAD

**Issue by t-muehlberger** [pocketbase/pocketbase#3759]
Hello, thank you for your great work on PocketBase, I used it for a little internal project and was pretty happy with how it went. I only have one issue. The process uses a lot of memory (over 1.5GB) when resizing thumbnails and eventually is OOM-Killed.

My tool lets people upload high resolution photos (many are over 5MB). When I open that PB-Dashboard to view the collection, the browser tries to load many thumbnails (40 per page) at the same time. This causes the server to use a lot of memory and eventually crash (usually at the second Page).  This behavior can only be observed on newly uploaded photos for which the thumbnails have not yet been generated. 

Now I understand that resizing large images requires that they are kept in memory and some memory usage is inevitable. But I am running the container with a 2GB memory limit and I think this should be sufficient. 

Also what seems a bit strange to me is that the memory usage is way above the actual size of the images. Resizing 80 Images with around 5MB each (400MB total) results in a memory consumption of over 2GB. 

**Possible Solution:** I think it would be really important to be able to limit the number of thumbnails that are generated concurrently. 

Keep in mind, that generating many many thumbnails at the same time will probably slow the server down causing the next thumbnails to be generated even slower and thus kept in memory even longer.

Also might be the case that the same thumbnail is generated more than once but I am not sure about that. I can Imagine that users refreshing the page when thumbnails are slow to load makes the problem even worse because the browser might be requesting the same thumbnail, that is already processing, again. 

It might also be possible that the concurrency is not the core issue here and that there is a memory leak somewhere but I don't know about that. Definitely should be further investigated, it is possible that limiting the concurrency would only hide some of the symptoms of a different problem.

Also, if you can point me into the right direction I can try to create a PR myself to fix the issue.

**Comment by ganigeorgiev** [comment]
If you are experiencing memory issues, you can try defining the [`GOMEMLIMIT=500MiB` env variable](https://pkg.go.dev/runtime) (_it is a soft limit and will instruct the GC to run more often once the threshold is reached_).

I don't think that there is a memory leak, but bulks thumb generation can definitely be improved.

We can try to create a worker pool and restrict the thumb generation processes. I'm currently busy with other tasks, but if anyone want to pick it the thumbs generation code is located in 
https://github.com/pocketbase/pocketbase/blob/985ab1e5b79a328b516decca3d1d7dc8edfcbba0/tools/filesystem/filesystem.go#L393.
It is invoked as part of the download endpoint in https://github.com/pocketbase/pocketbase/blob/985ab1e5b79a328b516decca3d1d7dc8edfcbba0/apis/file.go#L154.

**Comment by t-muehlberger** [comment]
I have already set GOMEMLIMIT without much success.

But I will try to have a look at the issue and maybe create a PR.

**Comment by ganigeorgiev** [comment]
Are you sure that the env variable is loaded by the container?

**Comment by t-muehlberger** [comment]
I will check the env variable again. 

S3 storage is used.

**Comment by ganigeorgiev** [comment]
> S3 storage is used.

In this case the memory consumption can be actually caused by the filesystem abstraction because we still need to reread and upload the generated thumb. It will need more elaborate investigation to pinpoint the exact cause of the issue (_note that the `gocloud.dev/blob` and `aws/aws-sdk-go` most likely will be removed in the future since they are currently ~50% of the binary size_).

**Comment by t-muehlberger** [comment]
Still I think limiting concurrency should probably improve the situation as fewer full-sized images would have to be read and kept in memory at the same time in the FS abstraction.

**Comment by ganigeorgiev** [comment]
Yes, definitely. Adding workers pool and somewhere to temp store the currently resizing thumb identifier will also prevent the edge case when multiple clients request a new thumb at the same time (_currently I guess it will generate and overwrite the same thumb for each request in this situation which is not ideal_).
If you have the time, feel free to open a PR with the suggested implementation. If no one picks it, after the logs generalization tasks I'll have a more detailed look at it.

**Comment by t-muehlberger** [comment]
I will try to have a look when i find the time in the next few days.

**Comment by aleBranching** [comment]
hi ganigeorgiev, just wondering, is this gonna be part of 0.23 main branch and if not do the original recommendation for extending still stand or would there be something else to consider. Might have a go at this but can't promise much. 

I had run into this issue myself when running pocketbase on a $5 Linode vps. with 1Gb ram.

**Comment by ganigeorgiev** [comment]
@aleBranching You are comment to a very old issue. The above mentioned concurrent thumb generation limit is already available in PocketBase v0.20+.

If you think that it can be done better feel free to open a PR and I'll try to review it but the note and the recommendations in my comment still remains.
